src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,308 @@
+        }
+
+        private static bool WithTransactionCommitProcessingWithRetry(IClientSessionHandle clientSession, DateTime startTime, IDateTimeProvider dateTimeProvider, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    clientSession.CommitTransaction(cancellationToken);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = dateTimeProvider.UtcNow;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !IsWriteConcernTimeoutError(mongoException) &&
+                            !HasTimedOut(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                        {
+                            // the transaction will be full processed once again.
+                            return false;
+                        }
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private static async Task<bool> WithTransactionCommitProcessingWithRetryAsync(IClientSessionHandle clientSession, DateTime startTime, IDateTimeProvider dateTimeProvider, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+
+                    await clientSession.CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    return false;
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = dateTimeProvider.UtcNow;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !IsWriteConcernTimeoutError(mongoException) &&
+                            !HasTimedOut(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                        {
+                            // the transaction will be full processed once again.
+                            return true;
+                        }
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        // nested types
+        internal class WithTransactionCallbackOutcome<TResult>
```

Comment:
Since this is a nested class and the scope of this class is limited we can use a shorter name:

    CallbackOutcome

Reply:
Done